### PR TITLE
Revamp RequestLog API

### PR DESCRIPTION
**Pull request by trustin**
Motivation:

RequestLog API currently notifies the log consumer only when a Request
or a Response is fully processed. This does not work well with a
streaming request that can take potentially long time when the consumer
wants to get certain information sooner.

A good example would be the implementation of 'activeRequest' counter.
DropwizardMetricConsumer increments the 'activeRequest' counter in
onRequest() handler method and decrements in onResponse(). This is OK
for Thrift calls where request processing takes small amount of time,
but not OK for other streaming calls such as file upload.

There should be a way to let a log consumer tell the API what properties
the log consumer is interested in and RequestLog should notify the
consumer as soon as the interested properties are available.

Modifications:

- Rewrite RequestLog and RequestLogBuilder
  - ResponseLog and ResponseLogBuilder are merged into RequestLog and
    RequestLogBuilder for simplicity
  - Add RequestLogAvailability to define the availability of properties
- Remove LogCollecting{Client,Service} because it doesn't do anything
  useful anymore with the new API
- Rename AbstractRpc{Request,Response} to DefaultRpc{Request,Response}
- Remove ThriftCall and ThriftReply and use DefaultRpc{Request,Response}
  - There was no reason to introduce these classes from the beginning...
- Rename ApacheThrift{Call,Reply} to Thrift{Call,Reply}
- Add the documentation about the new API
  - Update CSS a little bit to make things look better
  - Add a placeholder page for @kawamuray's Kafka structured logging
- Miscellaneous:
  - Make the formatter methods in TextFormatter append the raw value
    when the value is formatted into an approximated (or human-friendly)
    form

Result:

- Possible to get the necessary information sooner
- Cannot add an attribute to RequestLog anymore
  - May be both pro and con; will revisit if there's demand
- RequestLog API now has documentation.

### CODE REVIEW

**Comment by trustin** [comment]
Who's up for reviewing this PR? /cc @kawamuray @inch772 @imasahiro @anuraaga

**Comment by trustin** [comment]
Documentation preview:
http://armeria.readthedocs.io/en/revamp_request_log/advanced-structured-logging.html

**Comment by kojilin** [comment]
Curious about that do we need to provide more concrete http service class which contains `ctx.logBuilder().serializationFormat(SerializationFormat.NONE);` to avoid user forget to call ?

**Comment by trustin** [comment]
@kojilin I'm having hard time getting a reviewer for this PR. Are you interested in this?

**Comment by kojilin** [comment]
yeah, I'm start to reviewing.

**Comment by trustin** [comment]
@kojilin Thank you so much! Assigned to you. :-)

File: logback/src/main/java/com/linecorp/armeria/common/logback/RequestContextExportingAppender.java [hunk 1]
```diff
@@ -209,25 +209,16 @@ protected void append(ILoggingEvent eventObject) {
...
-            } else {
-                req = null;
-            }
-
-            if (ctx.responseLogFuture().isDone()) {
-                res = ctx.responseLogFuture().join();
-            } else {
-                res = null;
-            }
+            log = ctx.log();
```
Comment (kojilin) [inline logback/src/main/java/com/linecorp/armeria/common/logback/RequestContextExportingAppender.java]: join to previous line

File: core/src/main/java/com/linecorp/armeria/common/logging/DefaultRequestLog.java [hunk 2]
```diff
@@ -16,106 +16,720 @@
...
+    private boolean isAvailable(int interestedFlags) {
+        return (flags & interestedFlags) == interestedFlags;
+    }
+
+    private static boolean isAvailable(int flags, RequestLogAvailability availability) {
+        final int interestedFlags = availability.getterFlags();
+        return (flags & interestedFlags) == interestedFlags;
+    }
+
+    private boolean isAvailabilityUpdatable(RequestLogAvailability availability) {
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/common/logging/DefaultRequestLog.java]: You're right. Wrong name :-)

File: core/src/main/java/com/linecorp/armeria/common/logging/RequestLogListener.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+     * Returns a composed listener that calls this listener first and then the specified one.
+     */
+    default RequestLogListener andThen(RequestLogListener other) {
+        Objects.requireNonNull(other, "other");
+
+        final RequestLogListener first = this;
+        final RequestLogListener second = other;
+
+        return log -> {
+            RequestLogListenerInvoker.invokeOnRequestLog(first, log);
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/common/logging/RequestLogListener.java]: I'm not sure I understood your question. I wanted to make sure the second listener is always invoked even if the first one raised an exception.

File: logback/src/main/java/com/linecorp/armeria/common/logback/RequestContextExportingAppender.java (hunk 1 above)
Comment (trustin) [inline logback/src/main/java/com/linecorp/armeria/common/logback/RequestContextExportingAppender.java]: Fixed

**Comment by trustin** [comment]
> LGTM, only small nit. Maybe some one can help review too. 

I think it's ready to get merged. :-)

> And looks like current RequestLog can be apply to gRPC(even stream request)?

Yes, I hope so.

File: core/src/main/java/com/linecorp/armeria/common/logging/RequestLogListener.java (hunk 3 above)
Comment (kojilin) [inline core/src/main/java/com/linecorp/armeria/common/logging/RequestLogListener.java]: Ah I see. thx

**Comment by trustin** [comment]
Added a follow-up commit that changes the behavior of `RequestLogBuilder.serializationFormat(...)` and `AbstractHttpService` based on @kojilin's feed back. Please review again.

**Comment by kojilin** [comment]
LGTM

**Comment by trustin** [comment]
Thanks a lot for a swift review, @kojilin !

**Pull request merged**